File: 5 kyu/Best travel/solution.pl
package Solution;

use v5.30;
use warnings;
use List::Util qw(sum0);
use experimental qw(declared_refs refaliasing signatures);

use Exporter qw(import);
our @EXPORT_OK = qw(choose_best_sum);

sub choose_best_sum($max, $k, $dstr) {
    my @dists = split m/,/, $dstr =~ s{\s+}{}gr;
    my @combs = 
       sort {$a <=> $b }
       grep {$_ <= $max} 
       map {sum0($_->@*)} combinations(\@dists, $k)->@*;
    
    return -1 unless @combs;
    return $combs[-1];
}

sub combinations ($array, $k, $state = [], $result = []) {
    if ($k == 0) {
        push ($result->@*, $state);
        return $result;
    }

    my $length = scalar $array->@*;
    for (my $i = 0; $i < $length; ++$i) {
        combinations([$array->@[($i + 1) .. $length - 1]], $k - 1, [$state->@*, $array->[$i]], $result);
    }

    return $result;
}
_______________________________________
package Solution;

use 5.30.0;
use strict;
use warnings;
use Exporter qw(import);
our @EXPORT_OK = qw(choose_best_sum);

use List::Util qw(sum);
use experimental qw(signatures);

sub combinations ($list, $k) { # Trusty combination function.
    return map { [$_] } @$list if $k <= 1;
    
    my @combinations;
    for (my $i = 0; $i + $k <= @$list; ++$i) {
        my $current = $list->[$i];
        my $rest    = [@$list[($i + 1) .. $#$list]];
        push @combinations, [$current, @$_] for combinations($rest, $k - 1);
    }
    
    return @combinations;
}


sub choose_best_sum ($t, $n, $d_string) {
    my @d = split(', ', $d_string);
    
    my @sorted_distances = sort { $b <=> $a } map { sum(@$_) } combinations(\@d, $n);
    for (@sorted_distances) {
        return $_ if $_ <= $t;
    }
    
    return -1;
}
_______________________________________
package Solution;

use 5.30.0;
use strict;
use warnings;
use List::Util qw(max sum);
use Exporter qw(import);
our @EXPORT_OK = qw(choose_best_sum);

# parameter: maximum sum of distances, number of towns to visit, string of distances
sub choose_best_sum {
    my ($t, $k, $str) = @_;
    my $list = [map {int($_)} split(/,\s+/, $str)];
    return (max grep {$_ <= $t} map {sum @$_} combinations($list, $k)) || -1;
}

sub combinations {
    my ($listref, $n) = @_;
    return map {[$_]} @$listref if ($n <= 1);
    my @ret;
    for (my $i = 0; $i+$n <= @$listref; ++$i) {
        my $val  = $listref->[$i];
        my @rest = @$listref[$i+1..$#$listref];
        push(@ret, [$val, @$_]) for combinations(\@rest, $n-1);
    }
    return @ret;
}
_______________________________________
package Solution;

use 5.30.0;
use strict;
use warnings;
use Exporter qw(import);
our @EXPORT_OK = qw(choose_best_sum);

sub som {
    my $sum = 0;
    for ( @_ ) { $sum += $_; }
    return $sum;
}
sub combine {
  my ($list, $n) = @_;
  return map [$_], @$list if $n <= 1;
  my @comb;
  for (my $i = 0; $i+$n <= @$list; ++$i) {
    my $val  = $list->[$i];
    my @rest = @$list[$i+1..$#$list];
    push @comb, [$val, @$_] for combine(\@rest, $n-1);
  }
  return @comb;
}
sub aux {
  my ($t, $n, $list) = @_;
  my @res; my $mx = -1; my $r = -1;
  @res = map( som(@$_), @{$list});
  foreach (@res) {
      if (($_ > $mx) && ($_ <= $t)) { 
          $r = $_; $mx = $r; 
      }
  }
  return $r;
}
# parameter maximum sum of distances, number of towns to visit, string of distances
sub choose_best_sum {
    my ($t, $n, $s) = @_; 
    my @a = split /, /, $s;
    my @comb = combine(\@a, $n);
    return aux($t, $n, \@comb);
}

File: 5 kyu/Best travel/solution.py
import itertools
def choose_best_sum(t, k, ls):
    try: 
        return max(sum(i) for i in itertools.combinations(ls,k) if sum(i)<=t)
    except:
        return None
      
_______________________________________
from itertools import combinations

def choose_best_sum(t, k, ls):
    return max((s for s in (sum(dists) for dists in combinations(ls, k)) if s <= t), default=None)
_______________________________________
from itertools import combinations

def choose_best_sum(t, k, ls):
    return max((sum(v) for v in combinations(ls,k) if sum(v)<=t), default=None)
_______________________________________
def choose_best(t,k,ls):
    if k == 0: return 0
    best = -1
    for i, v in enumerate(ls):
        if v > t: continue
        b = choose_best(t - v, k - 1, ls[i+1:])
        if b < 0: continue
        b += v
        if b > best and b <= t:
            best = b
    return best

def choose_best_sum(t, k, ls):
    c = choose_best(t,k,ls)
    if c <= 0 : return None
    return c
_______________________________________
from itertools import combinations


def choose_best_sum(max_miles, max_towns, lst):
    highest = 0
    for a in combinations(lst, max_towns):
        total_distance = sum(a)
        if max_miles >= total_distance > highest:
            highest = total_distance
    return highest or None

_______________________________________
from itertools import combinations

def choose_best_sum(t, k, ls):
    return max([sum(comb)
                for comb in combinations(ls, k)
                if sum(comb) <= t] or [None])
_______________________________________
from itertools import combinations
from functools import reduce

def choose_best_sum(t, k, ls):
    mx = -1
    res = []
    for c in combinations(ls, k):
        s = reduce(lambda x, y: x + y, c)
        if ((s >= mx) and (s <= t)):
            res = [c, s]
            mx = s
    if (res == []): return None 
    else: return res[1]
_______________________________________
from itertools import combinations
def choose_best_sum(t, k, ls):
    return max((s for s in map(sum, combinations(ls,k)) if s <= t), default=None)
_______________________________________
from itertools import combinations

def choose_best_sum(max_distance, k, distances):
    best = 0
    for combination in combinations(distances, k):
        distance = sum(combination)
        if distance == max_distance:
            return distance
        if distance < max_distance:
            best = max(best, distance)
    return best if best > 0 else None
_______________________________________
def recurse(sum, ls, level):
  if level == 1:
    return [(x+sum) for x in ls]
  ary = list(ls)
  totals = []
  for x in ls:
    ary.remove(x)
    if len(ary) >= level - 1:
      totals += recurse(sum+x, ary, level - 1)
  return totals

def choose_best_sum(t, k, ls):
    if len(ls) < k:
      return None
    totals = recurse(0, ls, k)
    sum = 0
    for x in totals:
      if x > sum and x <= t:
        sum = x
    if sum == 0:
      return None
    return sum
_______________________________________
def choose_best_sum(t, k, ls):
    solutions = set()
    recursive_search(t, k, 0, ls, 0, 0, solutions)
    if len(solutions)>0:
        return max(solutions)
    else:
        return None
    
def recursive_search(t, maxk, k, ls, ind, solution, solutions):
    if ind == len(ls) or k==maxk or maxk - k > len(ls) - ind:
        return
    recursive_search(t, maxk, k, ls, ind+1, solution, solutions)
    k += 1
    solution += ls[ind]
    if solution <= t:
        if k == maxk:
            if solution in solutions:
                return
            solutions.add(solution)
            return
        recursive_search(t, maxk, k, ls, ind+1, solution, solutions)
_______________________________________
from itertools import combinations as comb

def choose_best_sum(t, k, ls):
    return reduce(lambda s, e: max(sum(e), s) if sum(e) <= t else s, comb(ls, k), None)
